### PR TITLE
#570 ThrowsConstraint failure message should include stack trace of actual exception

### LINKED ISSUE #570: ThrowsConstraint failure message should include stack trace of actual exception
When ThrowsConstraint fails due to the wrong exception Type being thrown, no info is given about where that exception was thrown. The message should include the stack trace of the exception in addition to the trace showing where the assert is located.

### PR DESCRIPTION
**Pull request by voloda**
Hi guys. Fixes #570. Feel free to provide me with additional guidance or simply reject.

### CODE REVIEW

**Comment by voloda** [comment]
Any suggestion how to fix the travis stack trace formatting difference?

File: src/NUnitFramework/framework/Constraints/ExceptionTypeConstraint.cs [hunk 1]
```diff
@@ -37,24 +37,47 @@ public class ExceptionTypeConstraint : ExactTypeConstraint
...
-        //    }
-        //}
+        /// <summary>
+        /// Applies the constraint to an actual value, returning a ConstraintResult.
+        /// </summary>
+        /// <param name="actual">The value to be tested</param>
+        /// <returns>A ConstraintResult</returns>
+        public override ConstraintResult ApplyTo<TActual>(TActual actual)
+        {
+            actualType = actual == null ? null : actual.GetType();
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Constraints/ExceptionTypeConstraint.cs]: What if actualType is not an Exception? Take a look at string constraints, for example, to see what we do when a constraint that expects a certain kind of actual value doesn't get it. Admittedly, it's a highly unlikely scenario requiring explciit construction of the constraint.

File: src/NUnitFramework/tests/Assertions/AssertThrowsTests.cs [hunk 2]
```diff
@@ -101,7 +101,7 @@ public void UnrelatedExceptionThrown()
             var ex = CatchException(() => Assert.Throws<ArgumentException>(TestDelegates.ThrowsNullReferenceException));
             Assert.That(ex.Message, Does.StartWith(
                 "  Expected: <System.ArgumentException>" + Env.NewLine +
-                "  But was:  <System.NullReferenceException>" + Env.NewLine));
+                "  But was:  <System.NullReferenceException: my message" + Env.NewLine + "   at NUnit.TestUtilities"));
```
Comment (CharliePoole) [inline src/NUnitFramework/tests/Assertions/AssertThrowsTests.cs]: You may run into cross-platform issues making assumptons about what's on top of the stack. We can wait and see, however.

**Comment by CharliePoole** [comment]
Aside from my inline comments, it looks good once you make it pass the Linux CI build.

**Comment by CharliePoole** [comment]
Assuming you have the eol characters right for the platform, I would drop the first part of the stack from the assert and just test the two lines. You can add another assert that tests the message contains "NUnit.TestUtilities" if you think it's needed.

A more thorough test would be to split the message into an array of lines and do a bunch of tests on them. I've done that in some parts of NUnit, but it's tedious and probably not need here.  BTW, the stack may not contain the word "at" in it if you're running under some other culture.

**Comment by voloda** [comment]
Thanks for feedback - I've switched for a while to PC with Czech locale enabled and there are more fragile unit tests (we use comma as decimal separator). I went with your proposal to check forEOL but not for the stack trace. In the end I am just re-using existing built-in behavior.

Also  I added verification of argument type to the constraint - just a comment - if you really compose it with Throws constraint it will most likely never happen since the passed in argument to composition is verified to be a delegate.

**Comment by voloda** [comment]
@CharliePoole if you have any other with similar complexity I am willing to help you. Just let me know.

**Comment by CharliePoole** [comment]
If you think that actual == null is not going to arise in context, we can leave it and merge.

@nunit/core-team  We need one other committer to approve the merge.

**Comment by CharliePoole** [comment]
@voloda There are currently four unassigned "easyfix" issues: https://github.com/nunit/nunit/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3Aeasyfix

**Comment by CharliePoole** [comment]
As soon as another committer reviews, we can merge this.

**Comment by rprouse** [comment]
Looks good to me, thanks for the contribution.

**Pull request merged**